app/src/main/java/com/bogdan/codeforceswatcher/features/problems/ProblemsFragment.kt

```diff
@@ -1,126 +1,389 @@
+                        keyboardOptions = KeyboardOptions(
+                            keyboardType = KeyboardType.Ascii,
+                            imeAction = ImeAction.Done
+                        ),
+                        keyboardActions = KeyboardActions(
+                            onDone = { localFocusManager.clearFocus() }
+                        ),
+                        onValueChange = { query ->
+                            problemsRequest(query)
+                        }
+                    )
+                }
+
+                Image(
+                    painter = painterResource(R.drawable.ic_search_icon),
+                    contentDescription = null,
+                    modifier = Modifier.clickable { isShownSearchTextFieldState = true }
+                )
+            }
+
+            if (isShownSearchTextFieldState) {
+                CrossButton(
+                    onCross = {
+                        isShownSearchTextFieldState = false
+                        problemsRequest("")
+                    }
+                )
+            } else {
+                FilterButton(
+                    onFilter = onFilter
+                )
+            }
+        }
     }
 }
+
+@Composable
+private fun CrossButton(
+    onCross: () -> Unit
+) = Image(
+    painter = painterResource(R.drawable.ic_cross_icon),
+    contentDescription = null,
+    modifier = Modifier.clickable { onCross() }
+)
+
+@Composable
+private fun FilterButton(
+    onFilter: () -> Unit
+) = Image(
+    painter = painterResource(R.drawable.ic_filter_icon),
+    contentDescription = null,
+    modifier = Modifier.clickable { onFilter() }
+)
+
+@Composable
+private fun SearchTextField(
+    modifier: Modifier = Modifier,
+    keyboardOptions: KeyboardOptions,
+    keyboardActions: KeyboardActions,
+    onValueChange: (String) -> Unit
+) {
+    var value by remember { mutableStateOf("") }
+
+    BasicTextField(
+        value = value,
+        onValueChange = {
+            value = it
+            onValueChange(it)
+        },
+        modifier = modifier
+            .height(30.dp)
+            .clip(RoundedCornerShape(20.dp))
+            .background(AlgoismeTheme.colors.primary)
+            .padding(10.dp, 4.dp, 0.dp, 0.dp),
+        textStyle = AlgoismeTheme.typography.primaryRegular.copy(color = AlgoismeTheme.colors.onBackground),
+        singleLine = true,
+        cursorBrush = SolidColor(AlgoismeTheme.colors.onBackground),
+        decorationBox = { innerTextField ->
+            if (value.isEmpty()) {
+                Text(
+                    text = "Search problems for...",
+                    style = AlgoismeTheme.typography.primaryRegular,
+                    color = AlgoismeTheme.colors.secondaryVariant
+                )
+            }
+            innerTextField()
+        },
+        keyboardOptions = keyboardOptions,
+        keyboardActions = keyboardActions
+    )
+}
+
+private data class UIModel(
```

Comment:
IMO, not needed.

Reply:
Why?

Reply:
It's not very common to use `UIModel` with JC. Usually, all values are just passed directly. Here, if you pass the state to the higher-level components, and then just values needed to lower-level, it should be OK.